library/src/main/java/lab/galaxy/yahfa/HookMain.java

```diff
@@ -76,78 +77,78 @@ private static void doHookItemDefault(ClassLoader patchClassLoader, String hookI
         }
     }
 
+    public static void findAndHook(Class targetClass, String methodName, String methodSig, Method hook) {
+        hook(findMethod(targetClass, methodName, methodSig), hook);
+    }
 
     public static void findAndBackupAndHook(Class targetClass, String methodName, String methodSig,
                                      Method hook, Method backup) {
-        try {
-            int hookParamCount = hook.getParameterTypes().length;
-            int targetParamCount = getParamCountFromSignature(methodSig);
-            Log.d(TAG, "target method param count is "+targetParamCount);
-            boolean isStatic = (hookParamCount == targetParamCount);
-            // virtual method has 'thiz' object as the first parameter
-            findAndBackupAndHook(targetClass, methodName, methodSig, isStatic, hook, backup);
+        backupAndHook(findMethod(targetClass, methodName, methodSig), hook, backup);
+    }
+
+    public static void hook(Method target, Method hook) {
+        backupAndHook(target, hook, null);
+    }
+
+    public static void backupAndHook(Method target, Method hook, Method backup) {
+        if (target == null) {
+            throw new IllegalArgumentException("null target method");
         }
-        catch (Exception e) {
-            e.printStackTrace();
+        if (hook == null) {
+            throw new IllegalArgumentException("null hook method");
         }
-    }
 
-    private static int getParamCountFromSignature(String signature) throws Exception{
-        int index;
-        int count = 0;
-        int seg;
-        try { // Read all declarations between for `(' and `)'
-            if (signature.charAt(0) != '(') {
-                throw new Exception("Invalid method signature: " + signature);
-            }
-            index = 1; // current string position
-            while(signature.charAt(index) != ')') {
-                seg = parseSignature(signature.substring(index));
-                index += seg;
-                count++;
-            }
+        checkCompatibleMethods(target, hook);
+        if (backup != null) {
+            checkCompatibleMethods(backup, target);
```

Comment:
@paulo-raca I guess it should be `checkCompatibleMethods(target, backup);` here, since the 2nd argument has to be a static method.